### PR TITLE
fix(gatsby): restore DEV SSR and make it work with getServerData

### PR DESCRIPTION
**Pull request by marvinjude**
## Description
DEV_SSR was disabled shortly before v4 release due to `getServerData` not being handled correctly. see - https://github.com/gatsbyjs/gatsby/pull/33054#discussion_r707189586

This PR allows enabling DEV_SSR and makes it work well with `getServerData`

## Documentation
DEV_SSR  - https://www.gatsbyjs.com/docs/reference/release-notes/v2.28/#feature-flags-in-gatsby-configjs

## Related Issues
[sc-49452]

## Todo
- [ ] Add e2e test for ssr page with `getServerData` and DEV_SSR flag

### CODE REVIEW

File: e2e-tests/development-runtime/gatsby-config.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ module.exports = {
     },
   },
   flags: {
-    DEV_SSR: false,
+    DEV_SSR: true,
```
Comment (pieh) [inline e2e-tests/development-runtime/gatsby-config.js:10]: We shouldn't do this here (not yet at least) - https://github.com/gatsbyjs/gatsby/tree/master/integration-tests/ssr are the test specifically for DEV_SSR

File: packages/gatsby/src/bootstrap/requires-writer.ts [hunk 2]
```diff
@@ -333,9 +332,8 @@ export const startListener = (): void => {
      * Start listening to CREATE_SERVER_VISITED_PAGE events so we can rewrite
      * files as required
      */
-    emitter.on(`CREATE_SERVER_VISITED_PAGE`, (): void => {
-      reporter.pendingActivity({ id: `requires-writer` })
-      debouncedWriteAll()
+    emitter.on(`CREATE_SERVER_VISITED_PAGE`, async (): void => {
+      await writeAll(store.getState())
     })
```
Comment (pieh) [inline packages/gatsby/src/bootstrap/requires-writer.ts:337]: this being debounced here cause whole lot of problems so I got rid of it - it sure is a trade, but timing issues caused by debounce felt worse to deal with than potentially writing out same file a lot in quick succession

File: integration-tests/ssr/__tests__/fixtures/bad-page.js [hunk 3]
```diff
@@ -2,6 +2,7 @@ import React from "react"
 
 const Component = () => {
   const a = window.width
+  console.log(a)
```
Comment (pieh) [inline integration-tests/ssr/__tests__/fixtures/bad-page.js:5]: Gnet rids of eslint warning about unused variable.

File: integration-tests/ssr/package.json [hunk 4]
```diff
@@ -40,7 +40,9 @@
     "develop": "cross-env GATSBY_EXPERIMENTAL_DEV_SSR=true CI=1 FORCE_COLOR=0 gatsby develop",
     "serve": "gatsby serve",
     "start-dev-server": "start-server-and-test develop http://localhost:8000 test:jest",
-    "test": "cross-env GATSBY_EXPERIMENTAL_DEV_SSR=true npm-run-all -s build start-dev-server",
+    "capture-prod-dsg-and-ssr": "start-server-and-test serve http://localhost:9000 write-dsg-and-ssr-to-files",
+    "write-dsg-and-ssr-to-files": "node write-dsg-and-ssr-to-files.js",
```
Comment (pieh) [inline integration-tests/ssr/package.json:44]: SSR page is not written to `public` (that's good :) ) but we need to have rawhtml from it, so those helpers run `gatsby serve` after the build and fetch dsg/ssr pages and save it to fs (gitignored `<rootOfSite>/prod-engines-output/<something>.html`) to compare later with DEV_SSR version of those pages.

For now we have single non-error SSR page to assert it works.

File: integration-tests/ssr/src/pages/usingtailwind.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
 import React from "react"
 import "../styles/tailwind.css"
 
-export default () => <h1 className="text-3xl">This is a 3xl text</h1>
+export default function PageUsingTailwind() {
+  return <h1 className="text-3xl">This is a 3xl text</h1>
+}
```
Comment (pieh) [inline integration-tests/ssr/src/pages/usingtailwind.js:6]: name the default export so fash-refresh is happy

File: packages/gatsby/cache-dir/ssr-develop-static-entry.js [hunk 6]
```diff
@@ -237,7 +241,7 @@ export default async function staticPage(
           !isClientOnlyPage
         ) {
           pageElement = React.createElement(
-            syncRequires.ssrComponents[componentChunkName],
+            preferDefault(syncRequires.ssrComponents[componentChunkName]),
```
Comment (pieh) [inline packages/gatsby/cache-dir/ssr-develop-static-entry.js:244]: `preferDefault` removed from `ssr-sync-requires` (to allow bundling and executing user's `getServerData`) so we use it here

File: packages/gatsby/src/commands/build-html.ts [hunk 7]
```diff
@@ -83,18 +145,22 @@ const runWebpack = (
...
       })
 
+      let isFirstCompile = true
       const watcher = compiler.watch(
         {
           ignored: /node_modules/,
         },
         (err, stats) => {
           // this runs multiple times
-          needToRecompileSSRBundle = false
```
Comment (pieh) [inline packages/gatsby/src/commands/build-html.ts:95]: setting it to `false` here was possibly resetting any `invalid` events that happen during compilation, so this was moved to place where recompilation is triggered

File: packages/gatsby/src/utils/dev-ssr/render-dev-html.ts [hunk 8]
```diff
@@ -3,15 +3,17 @@ import fs from "fs-extra"
...
 import { findPageByPath } from "../find-page-by-path"
 import { getPageData as getPageDataExperimental } from "../get-page-data"
 import { getDevSSRWebpack } from "../../commands/build-html"
-import { emitter, GatsbyReduxStore } from "../../redux"
+import { GatsbyReduxStore } from "../../redux"
 import { IGatsbyPage } from "../../redux/types"
-
+import { getServerData, IServerData } from "../get-server-data"
+import { getPageMode } from "../page-mode"
+import { parseError } from "./render-dev-html-child"
```
Comment (pieh) [inline packages/gatsby/src/utils/dev-ssr/render-dev-html.ts:16]: this probably should be moved to some common module and not imported from worker / child process entry

File: packages/gatsby/src/utils/dev-ssr/render-dev-html.ts [hunk 9]
```diff
@@ -22,7 +24,8 @@ interface IErrorRenderMeta {
 }
 
 // TODO: convert `render-dev-html-child.js` to TS and use `typeof import("./render-dev-html-child")`
-// instead of defining interface here
+// instead of defining interface hereimport { IServerData } from '../../../dist/utils/get-server-data.d';
+
```
Comment (pieh) [inline packages/gatsby/src/utils/dev-ssr/render-dev-html.ts:28]: TODO: revert this change

File: packages/gatsby/cache-dir/ssr-develop-static-entry.js [hunk 10]
```diff
@@ -48,13 +51,13 @@ try {
 
 Html = Html && Html.__esModule ? Html.default : Html
 
-export default async function staticPage(
+export default async function staticPage({
```
Comment (LekoArts) [inline packages/gatsby/cache-dir/ssr-develop-static-entry.js:54]: 👍 to moving to object syntax here

File: packages/gatsby/src/commands/build-html.ts [hunk 11]
```diff
@@ -46,20 +45,83 @@ interface IBuildRendererResult {
...
+          emitter.on(`DEV_SSR_COMPILATION_DONE`, finish)
+          // we reset it before we start compilation to be able to catch any invalid events
+          SSRBundleReceivedInvalidEvent = false
+          if (activeRecompilations === 0) {
+            watcher.resume()
+          }
+          activeRecompilations++
+
+          if (allowTimedFallback) {
+            // Timeout after 1.5s.
```
Comment (LekoArts) [inline packages/gatsby/src/commands/build-html.ts:106]: I know we had this the same previously (and it's probably arbitrary). Any good reasons to keep it as-is or increase/decrease it? I'm fine with keeping it like that

File: packages/gatsby/src/utils/dev-ssr/parse-error.ts [hunk 12]
```diff
@@ -4,7 +4,11 @@ const sysPath = require(`path`)
 const fs = require(`fs-extra`)
 const { slash } = require(`gatsby-core-utils`)
```
Comment (LekoArts) [inline packages/gatsby/src/utils/dev-ssr/parse-error.ts:6]: The TODO here

```
// TODO: Do not use source-map-support and migrate to a package that doesn't use source-map under the hood
```

can be solved with: https://github.com/cspotcode/node-source-map-support

**Review by LekoArts — COMMENTED** [review]
Overall LGTM, the TODO needs to be handled for Node 18 support

File: packages/gatsby/src/commands/build-html.ts (hunk 11 above)
Comment (pieh) [reply packages/gatsby/src/commands/build-html.ts:106]: I decided not to change any of timeouts we already have in this PR - the goal was only to make it work again at all here, and as I mentioned couple of time - there is a lot of opporunities to improve the way DEV_SSR works, I just didn't want this PR to get stuck on those questions, as improvements should be a follow up to changes done here which are mostly fixes

File: packages/gatsby/src/utils/dev-ssr/parse-error.ts (hunk 12 above)
Comment (pieh) [reply packages/gatsby/src/utils/dev-ssr/parse-error.ts:6]: I removed the usage of ``require(`source-map-support`).install()`` and decided not to use alternative that also do some "magic installation" and instead explicitely use `createErrorFromString` we already have that converts stack trace to use source call sites in https://github.com/gatsbyjs/gatsby/pull/35685/commits/a1b033ae64a125cfbf12d09ba82972e4c75a4781
